File: apps/daemon/src/router/handlers/files.rs
//! File operation handlers

use std::sync::Arc;

use axum::{
    extract::Query,
    Extension, Json,
};
use serde::{Deserialize, Serialize};
use tracing::{debug, info, warn};

use crate::filesystem::{FileInfo, Filesystem};
use crate::server::Server;
use super::ApiError;

// url crate re-used from Cargo.toml dependency

/// List files request
#[derive(Debug, Deserialize)]
pub struct ListFilesQuery {
    #[serde(default)]
    pub directory: String,
}

/// List files in a directory
pub async fn list_files(
    Extension(server): Extension<Arc<Server>>,
    Query(query): Query<ListFilesQuery>,
) -> Result<Json<Vec<FileInfo>>, ApiError> {
    let fs = get_filesystem(&server)?;
    let files = fs.list_directory(&query.directory).await?;
    Ok(Json(files))
}

/// Read file request
#[derive(Debug, Deserialize)]
pub struct ReadFileQuery {
    pub file: String,
}

/// Read file contents
pub async fn read_file(
    Extension(server): Extension<Arc<Server>>,
    Query(query): Query<ReadFileQuery>,
) -> Result<String, ApiError> {
    let fs = get_filesystem(&server)?;
    let contents = fs.read_file_string(&query.file).await?;
    Ok(contents)
}

/// Write file request
#[derive(Debug, Deserialize)]
pub struct WriteFileRequest {
    pub file: String,
    pub content: String,
}

/// Write file contents
pub async fn write_file(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<WriteFileRequest>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let fs = get_filesystem(&server)?;
    fs.write_file(&request.file, request.content.as_bytes()).await?;

    Ok(Json(serde_json::json!({
        "success": true
    })))
}

/// Create file or directory request
#[derive(Debug, Deserialize)]
pub struct CreateFileRequest {
    pub path: String,
    #[serde(rename = "type")]
    pub file_type: String,
    #[serde(default)]
    pub content: Option<String>,
}

/// Create a file or directory
pub async fn create_file(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<CreateFileRequest>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let fs = get_filesystem(&server)?;

    match request.file_type.as_str() {
        "directory" => {
            fs.create_directory(&request.path).await?;
            Ok(Json(serde_json::json!({
                "success": true
            })))
        }
        "file" => {
            let content = request.content.as_deref().unwrap_or("");
            fs.write_file(&request.path, content.as_bytes()).await?;
            Ok(Json(serde_json::json!({
                "success": true
            })))
        }
        _ => Err(ApiError::bad_request(format!(
            "Invalid type '{}'. Must be 'file' or 'directory'",
            request.file_type
        ))),
    }
}

/// Create directory request
#[derive(Debug, Deserialize)]
pub struct CreateDirectoryRequest {
    pub name: String,
    #[serde(default)]
    pub root: String,
}

/// Create a directory
pub async fn create_directory(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<CreateDirectoryRequest>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let fs = get_filesystem(&server)?;

    let path = if request.root.is_empty() {
        request.name
    } else {
        format!("{}/{}", request.root.trim_end_matches('/'), request.name)
    };

    fs.create_directory(&path).await?;

    Ok(Json(serde_json::json!({
        "success": true
    })))
}

/// Rename file request
#[derive(Debug, Deserialize)]
pub struct RenameFileRequest {
    pub from: String,
    pub to: String,
    #[serde(default)]
    pub root: String,
}

/// Rename a file or directory
pub async fn rename_file(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<RenameFileRequest>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let fs = get_filesystem(&server)?;

    let from = if request.root.is_empty() {
        request.from
    } else {
        format!("{}/{}", request.root.trim_end_matches('/'), request.from)
    };

    let to = if request.root.is_empty() {
        request.to
    } else {
        format!("{}/{}", request.root.trim_end_matches('/'), request.to)
    };

    fs.rename(&from, &to).await?;

    Ok(Json(serde_json::json!({
        "success": true
    })))
}

/// Copy file request
#[derive(Debug, Deserialize)]
pub struct CopyFileRequest {
    pub location: String,
}

/// Copy a file or directory
pub async fn copy_file(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<CopyFileRequest>,
) -> Result<Json<CopyFileResponse>, ApiError> {
    let fs = get_filesystem(&server)?;
    let new_name = fs.copy(&request.location).await?;

    Ok(Json(CopyFileResponse { name: new_name }))
}

#[derive(Debug, Serialize)]
pub struct CopyFileResponse {
    pub name: String,
}

/// Delete files request
#[derive(Debug, Deserialize)]
pub struct DeleteFilesRequest {
    pub files: Vec<String>,
    #[serde(default)]
    pub root: String,
}

/// Delete files or directories
pub async fn delete_files(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<DeleteFilesRequest>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let fs = get_filesystem(&server)?;

    for file in request.files {
        let path = if request.root.is_empty() {
            file
        } else {
            format!("{}/{}", request.root.trim_end_matches('/'), file)
        };

        fs.delete(&path).await?;
    }

    Ok(Json(serde_json::json!({
        "success": true
    })))
}

/// Compress files request
#[derive(Debug, Deserialize)]
pub struct CompressFilesRequest {
    pub files: Vec<String>,
    #[serde(default)]
    pub root: String,
}

/// Compress files into an archive
pub async fn compress_files(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<CompressFilesRequest>,
) -> Result<Json<FileInfo>, ApiError> {
    let fs = get_filesystem(&server)?;

    let root = if request.root.is_empty() {
        ".".to_string()
    } else {
        request.root
    };

    let info = fs.compress(&root, request.files).await?;

    Ok(Json(info))
}

/// Decompress file request
#[derive(Debug, Deserialize)]
pub struct DecompressFileRequest {
    pub file: String,
    #[serde(default)]
    pub root: String,
}

/// Decompress an archive
pub async fn decompress_file(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<DecompressFileRequest>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let fs = get_filesystem(&server)?;

    let destination = if request.root.is_empty() {
        ".".to_string()
    } else {
        request.root
    };

    fs.decompress(&request.file, &destination).await?;

    Ok(Json(serde_json::json!({
        "success": true
    })))
}

/// Chmod request
#[derive(Debug, Deserialize)]
pub struct ChmodRequest {
    pub files: Vec<ChmodEntry>,
    #[serde(default)]
    pub root: String,
}

#[derive(Debug, Deserialize)]
pub struct ChmodEntry {
    pub file: String,
    pub mode: u32,
}

/// Change file permissions
pub async fn chmod_file(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<ChmodRequest>,
) -> Result<Json<serde_json::Value>, ApiError> {
    let fs = get_filesystem(&server)?;

    for entry in request.files {
        let path = if request.root.is_empty() {
            entry.file
        } else {
            format!("{}/{}", request.root.trim_end_matches('/'), entry.file)
        };

        fs.chmod(&path, entry.mode).await?;
    }

    Ok(Json(serde_json::json!({
        "success": true
    })))
}

/// Pull file from URL request
#[derive(Debug, Deserialize)]
pub struct PullFileRequest {
    /// The remote URL to download from
    pub url: String,
    /// Target directory (relative to server root)
    #[serde(default)]
    pub directory: String,
    /// Optional filename override; if not set, derived from URL or Content-Disposition
    pub filename: Option<String>,
    /// Whether to decompress the downloaded file (supports .tar.gz, .zip, .tar)
    #[serde(default)]
    pub decompress: bool,
}

#[derive(Debug, Serialize)]
pub struct PullFileResponse {
    pub success: bool,
    pub path: String,
    pub size: u64,
}

/// Download a file from a remote URL into the server's filesystem
pub async fn pull_file(
    Extension(server): Extension<Arc<Server>>,
    Json(request): Json<PullFileRequest>,
) -> Result<Json<PullFileResponse>, ApiError> {
    // Validate URL
    let url: url::Url = request.url.parse().map_err(|e| {
        ApiError::bad_request(format!("Invalid URL: {}", e))
    })?;

    // Only allow http/https
    match url.scheme() {
        "http" | "https" => {}
        scheme => {
            return Err(ApiError::bad_request(format!(
                "Unsupported URL scheme '{}'. Only http and https are allowed.",
                scheme
            )));
        }
    }

    info!("Pulling file from URL: {} into directory: {}", url, request.directory);

    // Perform the download
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::limited(10))
        .timeout(std::time::Duration::from_secs(600))
        .build()
        .map_err(|e| ApiError::internal(format!("Failed to create HTTP client: {}", e)))?;

    let response = client.get(url.clone()).send().await.map_err(|e| {
        ApiError::internal(format!("Failed to download file: {}", e))
    })?;

    if !response.status().is_success() {
        return Err(ApiError::internal(format!(
            "Download failed with status: {}",
            response.status()
        )));
    }

    // Determine filename
    let filename = request.filename.unwrap_or_else(|| {
        // Try Content-Disposition header first
        if let Some(cd) = response.headers().get(reqwest::header::CONTENT_DISPOSITION) {
            if let Ok(cd_str) = cd.to_str() {
                // Parse filename from Content-Disposition: attachment; filename="foo.jar"
                if let Some(idx) = cd_str.find("filename=") {
                    let name = &cd_str[idx + 9..];
                    let name = name.trim_matches('"').trim_matches('\'');
                    if !name.is_empty() {
                        return name.to_string();
                    }
                }
            }
        }

        // Fall back to last URL path segment
        url.path_segments()
            .and_then(|segments| segments.last())
            .filter(|s| !s.is_empty())
            .map(|s| s.to_string())
            .unwrap_or_else(|| "download".to_string())
    });

    // Build target path
    let target_path = if request.directory.is_empty() {
        filename.clone()
    } else {
        format!("{}/{}", request.directory.trim_end_matches('/'), filename)
    };

    // Read response body
    let bytes = response.bytes().await.map_err(|e| {
        ApiError::internal(format!("Failed to read download response: {}", e))
    })?;
    let size = bytes.len() as u64;

    let fs = get_filesystem(&server)?;

    // Write the file
    fs.write_file(&target_path, &bytes).await?;

    info!("Downloaded {} bytes to {}", size, target_path);

    // Optionally decompress
    if request.decompress {
        let dest_dir = if request.directory.is_empty() {
            ".".to_string()
        } else {
            request.directory.clone()
        };

        info!("Decompressing {} to {}", target_path, dest_dir);
        fs.decompress(&target_path, &dest_dir).await?;

        // Delete the archive after extraction
        fs.delete(&target_path).await?;
        info!("Decompressed and cleaned up archive {}", target_path);
    }

    Ok(Json(PullFileResponse {
        success: true,
        path: target_path,
        size,
    }))
}

/// Helper to get filesystem for a server
fn get_filesystem(server: &Server) -> Result<Filesystem, ApiError> {
    let config = server.config();

    Filesystem::new(
        server.data_dir().clone(),
        config.disk_bytes(),
        config.egg.file_denylist.clone(),
    ).map_err(|e| ApiError::internal(e.to_string()))
}

/// Disk usage response
#[derive(Debug, Serialize)]
pub struct DiskUsageResponse {
    pub used_bytes: u64,
    pub limit_bytes: u64,
    pub path: String,
}

/// Get disk usage for the server
pub async fn disk_usage(
    Extension(server): Extension<Arc<Server>>,
) -> Result<Json<DiskUsageResponse>, ApiError> {
    let fs = get_filesystem(&server)?;
    let data_dir = server.data_dir();

    // Calculate actual disk usage
    let used = match fs.disk_usage().calculate(&data_dir).await {
        Ok(size) => {
            debug!("Successfully calculated disk usage for {:?}: {} bytes", data_dir, size);
            size
        }
        Err(e) => {
            warn!("Failed to calculate disk usage for {:?}: {}", data_dir, e);
            0
        }
    };

    let limit = server.config().disk_bytes();

    Ok(Json(DiskUsageResponse {
        used_bytes: used,
        limit_bytes: limit,
        path: "/".to_string(),
    }))
}

File: apps/mock-daemon/src/routes/Files.ts
/**
 * File operation routes for the mock daemon.
 * Implements list, read, write, create, rename, copy, delete, compress,
 * decompress, chmod, disk-usage, and pull endpoints.
 */

import { Hono } from "hono";
import type {
  WriteFileRequest,
  CreateFileRequest,
  RenameFileRequest,
  CopyFileRequest,
  DeleteFilesRequest,
  CompressFilesRequest,
  DecompressFileRequest,
  ChmodRequest,
  PullFileRequest,
  MockFile,
} from "../Types.js";
import { GetServer, ListFiles, GetFile, SetFile, DeleteFile, NormalizePath } from "../State.js";
import { MockFileToFileInfo, GenerateUUID } from "../Generators.js";

const FileRoutes = new Hono();

/**
 * GET /files/list?directory=... — List files in a directory.
 */
FileRoutes.get("/list", (c) => {
  const serverId = c.req.param("server_id") as string;
  const directory = c.req.query("directory") ?? "/";
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const files = ListFiles(serverId, directory);
  return c.json(files.map(MockFileToFileInfo));
});

/**
 * GET /files/contents?file=... — Read file contents.
 */
FileRoutes.get("/contents", (c) => {
  const serverId = c.req.param("server_id") as string;
  const filePath = c.req.query("file") ?? "";
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const file = GetFile(serverId, filePath);
  if (!file || file.is_directory) {
    return c.json({ error: "File not found" }, 404);
  }

  return c.text(file.content);
});

/**
 * POST /files/write — Write content to a file.
 */
FileRoutes.post("/write", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const body = await c.req.json<WriteFileRequest>();
  const now = Math.floor(Date.now() / 1000);
  const normalized = NormalizePath(body.file);
  const name = normalized.split("/").pop() ?? body.file;

  const existing = GetFile(serverId, body.file);
  const file: MockFile = {
    name,
    content: body.content,
    is_directory: false,
    size: body.content.length,
    modified: now,
    created: existing?.created ?? now,
    mode: existing?.mode ?? 0o644,
  };

  SetFile(serverId, body.file, file);
  return c.json({ success: true });
});

/**
 * POST /files/create — Create a new file or directory.
 */
FileRoutes.post("/create", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const body = await c.req.json<CreateFileRequest>();
  const now = Math.floor(Date.now() / 1000);
  const normalized = NormalizePath(body.path);
  const name = normalized.split("/").pop() ?? body.path;
  const isDir = body.type === "directory";

  const file: MockFile = {
    name,
    content: isDir ? "" : (body.content ?? ""),
    is_directory: isDir,
    size: isDir ? 4096 : (body.content?.length ?? 0),
    modified: now,
    created: now,
    mode: isDir ? 0o755 : 0o644,
  };

  SetFile(serverId, body.path, file);
  return c.json({ success: true });
});

/**
 * POST /files/create-directory — Create a new directory.
 */
FileRoutes.post("/create-directory", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const body = await c.req.json<{ name: string; path?: string }>();
  const now = Math.floor(Date.now() / 1000);
  const fullPath = body.path ? `${body.path}/${body.name}` : `/${body.name}`;

  const dir: MockFile = {
    name: body.name,
    content: "",
    is_directory: true,
    size: 4096,
    modified: now,
    created: now,
    mode: 0o755,
  };

  SetFile(serverId, fullPath, dir);
  return c.json({ success: true });
});

/**
 * POST /files/rename — Rename files or directories.
 */
FileRoutes.post("/rename", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const body = await c.req.json<RenameFileRequest>();
  const root = body.root ?? "/";

  for (const rename of body.files) {
    const fromPath = `${root}/${rename.from}`.replace(/\/+/g, "/");
    const toPath = `${root}/${rename.to}`.replace(/\/+/g, "/");
    const file = GetFile(serverId, fromPath);

    if (file) {
      const newName = toPath.split("/").pop() ?? rename.to;
      DeleteFile(serverId, fromPath);
      SetFile(serverId, toPath, { ...file, name: newName, modified: Math.floor(Date.now() / 1000) });
    }
  }

  return c.json({ success: true });
});

/**
 * POST /files/copy — Copy a file.
 */
FileRoutes.post("/copy", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const body = await c.req.json<CopyFileRequest>();
  const file = GetFile(serverId, body.location);

  if (!file) {
    return c.json({ error: "File not found" }, 404);
  }

  const now = Math.floor(Date.now() / 1000);
  const ext = file.name.includes(".") ? `.${file.name.split(".").pop()}` : "";
  const base = file.name.replace(ext, "");
  const copyName = `${base}-copy${ext}`;
  const dir = NormalizePath(body.location).split("/").slice(0, -1).join("/") || "/";
  const copyPath = `${dir}/${copyName}`;

  SetFile(serverId, copyPath, { ...file, name: copyName, created: now, modified: now });
  return c.json({ name: copyName });
});

/**
 * DELETE /files/delete — Delete files and directories.
 */
FileRoutes.delete("/delete", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const body = await c.req.json<DeleteFilesRequest>();
  const root = body.root ?? "/";

  for (const filePath of body.files) {
    const fullPath = `${root}/${filePath}`.replace(/\/+/g, "/");
    DeleteFile(serverId, fullPath);
  }

  return c.json({ success: true });
});

/**
 * POST /files/compress — Compress files into an archive.
 */
FileRoutes.post("/compress", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const body = await c.req.json<CompressFilesRequest>();
  const now = Math.floor(Date.now() / 1000);
  const archiveName = `archive-${GenerateUUID().slice(0, 8)}.tar.gz`;
  const archivePath = `${body.root}/${archiveName}`.replace(/\/+/g, "/");

  const archive: MockFile = {
    name: archiveName,
    content: `[compressed: ${body.files.join(", ")}]`,
    is_directory: false,
    size: 1024 * 1024,
    modified: now,
    created: now,
    mode: 0o644,
  };

  SetFile(serverId, archivePath, archive);
  return c.json(MockFileToFileInfo(archive));
});

/**
 * POST /files/decompress — Decompress an archive.
 */
FileRoutes.post("/decompress", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  await c.req.json<DecompressFileRequest>();
  return c.json({ success: true });
});

/**
 * POST /files/chmod — Change file permissions.
 */
FileRoutes.post("/chmod", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const body = await c.req.json<ChmodRequest>();
  const root = body.root ?? "/";

  for (const entry of body.files) {
    const fullPath = `${root}/${entry.file}`.replace(/\/+/g, "/");
    const file = GetFile(serverId, fullPath);
    if (file) {
      SetFile(serverId, fullPath, { ...file, mode: entry.mode });
    }
  }

  return c.json({ success: true });
});

/**
 * GET /files/disk-usage — Get disk usage for a server.
 */
FileRoutes.get("/disk-usage", (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  /** Sum up all file sizes */
  let used = 0;
  for (const file of server.files.values()) {
    used += file.size;
  }

  const total = (server.config.build?.disk_space ?? 10240) * 1024 * 1024;
  return c.json({ total, used, available: total - used });
});

/**
 * POST /files/pull — Pull a remote file to the server.
 */
FileRoutes.post("/pull", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const body = await c.req.json<PullFileRequest>();
  const now = Math.floor(Date.now() / 1000);
  const fileName = body.file_name ?? body.url.split("/").pop() ?? "downloaded-file";
  const dir = body.directory ?? "/";
  const fullPath = `${dir}/${fileName}`.replace(/\/+/g, "/");

  const file: MockFile = {
    name: fileName,
    content: `[pulled from: ${body.url}]`,
    is_directory: false,
    size: 1024,
    modified: now,
    created: now,
    mode: 0o644,
  };

  SetFile(serverId, fullPath, file);
  return c.json({ success: true });
});

/**
 * POST /files/upload — Handle file upload (multipart).
 */
FileRoutes.post("/upload", async (c) => {
  const serverId = c.req.param("server_id") as string;
  const server = GetServer(serverId);

  if (!server) {
    return c.json({ error: "Server not found" }, 404);
  }

  const directory = c.req.query("directory") ?? "/";
  const now = Math.floor(Date.now() / 1000);

  /** In mock mode, just acknowledge the upload */
  const uploadedFiles: string[] = [];
  try {
    const formData = await c.req.formData();
    for (const [key, value] of formData.entries()) {
      if (value instanceof File) {
        const fileName = value.name || key;
        const fullPath = `${directory}/${fileName}`.replace(/\/+/g, "/");
        const content = await value.text();

        SetFile(serverId, fullPath, {
          name: fileName,
          content,
          is_directory: false,
          size: value.size,
          modified: now,
          created: now,
          mode: 0o644,
        });
        uploadedFiles.push(fileName);
      }
    }
  } catch {
    /** If form parsing fails, still return success (mock) */
    uploadedFiles.push("mock-uploaded-file");
  }

  return c.json({ success: true, files: uploadedFiles });
});

export default FileRoutes;
